homeassistant/components/demo/event.py

```diff
@@ -0,0 +1,46 @@
+"""Demo platform that offers a fake event entity."""
+from __future__ import annotations
+
+from homeassistant.components.event import EventDeviceClass, EventEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import DOMAIN
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the demo event platform."""
+    async_add_entities([DemoEvent()])
+
+
+class DemoEvent(EventEntity):
+    """Representation of a demo event entity."""
+
+    _attr_device_class = EventDeviceClass.BUTTON
+    _attr_event_types = ["pressed"]
+    _attr_has_entity_name = True
+    _attr_name = "Button pushed"
+    _attr_should_poll = False
+    _attr_translation_key = "push"
+    _attr_unique_id = "push"
+
+    def __init__(self) -> None:
+        """Initialize the Demo event entity."""
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, "push")},
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        self.hass.bus.async_listen("demo_button_pressed", self._async_handle_event)
```

Comment:
We probably need a helper that subscribes to the events and dispatches only matching ones....a more generic version of `async_track_state_change_event`. That would avoid the pattern were we dispatch the event to potential listeners and they reject it because it doesn't match the relevant device etc.

Maybe we could make the event required to have the `device_id` in it so you only subscribe to events for the device id. It would make the events a bit more standard

Reply:
Logbook already knows how to handle the device_id in events so it would probably be a good idea to formalize something like that at this time

Reply:
Maybe the subscription tree looks something like `[event_type][(device_id, event_type)]`...not sure all the use cases here though

Reply:
So, maybe this example isn't the best to demonstrate implementation details.

I would not expect entities to listen to existing HA events. I would expect them to tie into whatever provides those events in the first place (provided by the package/library the integration uses).